### ISSUE TITLE
[Triton 3.7.1] Inductor inference latency regression on PVC — 50 models affected (10–20x slower)

### ISSUE THREAD

**Issue by mengfei25** [intel/intel-xpu-backend-for-triton#6992]
## Summary

Triton 3.7.1 introduces severe inductor latency regressions on **PVC (Ponte Vecchio)** for **inference-only** workloads. 50 model/dtype/mode combinations are affected, with inductor latency increasing **10–20x** compared to Triton 3.7.0 baseline.

**Training workloads are unaffected.** Eager latency is stable between target and baseline, confirming this is a Triton compilation/codegen issue, not an environment problem.

## Environment

| Component | Target | Baseline |
|-----------|--------|----------|
| Triton | 3.7.1 (`2bafaf80`) | 3.7.1 (`21033c4e2be`) |
| PyTorch | torch 2.12.0+xpu (nightly) | torch 2.12.0+xpu (nightly) |
| Platform | PVC (Data Center Max) | PVC (Data Center Max) |

## Affected Models

All are **inference mode** across multiple dtypes (amp_bf16, amp_fp16, bfloat16, float16).
Ratio = baseline_latency / target_latency (values < 1.0 mean target is slower):

| Model | Suite | Inductor Ratio | Eager Ratio |
|-------|-------|----------------|-------------|
| fbnetc_100 | timm_models | **0.054x** | 1.00x |
| ghostnet_100 | timm_models | **0.062x** | 1.00x |
| mnasnet_100 | timm_models | **0.055x** | 1.00x |
| gluon_inception_v3 | timm_models | **0.066x** | 1.03x |
| inception_v3 | timm_models | **0.065x** | 1.03x |
| res2net50_14w_8s | timm_models | **0.063x** | 1.00x |
| resnest101e | timm_models | **0.065x** | 1.00x |
| timm_resnest | torchbench | **0.065x** | 1.01x |
| botnet26t_256 | timm_models | **0.062x** | 1.00x |
| resnet152 | torchbench | **0.069x** | 1.00x |
| resnet50 | torchbench | **0.063x** | 1.00x |
| resnext50_32x4d | torchbench | **0.069x** | 1.00x |
| densenet121 | torchbench | **0.079x** | 1.00x |
| mobilenet_v2 | torchbench | **0.051x** | 1.00x |
| mobilenetv2_100 | timm_models | **0.050x** | 1.00x |
| mnasnet1_0 | torchbench | **0.057x** | 1.00x |
| spnasnet_100 | timm_models | **0.055x** | 1.00x |

## Verification

- **Eager latency ratio is ~1.0x** for all affected models, confirming the test environment is consistent.
- **Training mode is unaffected** — same models in training show inductor latency ratio within 0.97–1.03x.
- Results reproduced on rerun.

## Observations

- All affected architectures use **depthwise separable convolutions** or **mobile-style inverted residual blocks** (MobileNet family, EfficientNet family, ResNeSt, GhostNet).
- The regression appears to be in Triton kernel codegen or scheduling for these specific graph patterns during inference.
- The 10-20x regression suggests compilation is falling back to a suboptimal code path or generating excessive kernel launches.

## Steps to Reproduce

```bash
docker run -it -e TZ=Aisa/Shanghai --device=/dev/mem --device=/dev/dri --group-add video \
    --privileged --shm-size=8g intelgpu/ubuntu-24.04-lts2:2523.40 bash

# setup python
curl -LsSf https://astral.sh/uv/install.sh | env UV_INSTALL_DIR="/usr/local/bin" sh
uv venv myvenv --python 3.12 --clear
source myvenv/bin/activate
uv pip install pip numpy 'setuptools<81' wheel

# install pytorch nightly xpu wheel
uv pip install torch torchaudio torchvision --pre --index-url https://download.pytorch.org/whl/xpu

# prepare pytorch source
pytorch_commit="$(python -c 'import torch; print(torch.version.git_version)')"
git clone https://github.com/pytorch/pytorch
cd pytorch
git checkout ${pytorch_commit}

# install deps
uv pip install pip pandas psutil scipy requests
uv pip install -U numpy==1.26.4 accelerate
uv pip install -r .ci/docker/ci_commit_pins/huggingface-requirements.txt
uv pip install timm

# sync model list
rm -rf torch torch-xpu-ops
git clone https://github.com/intel/torch-xpu-ops
rsync -avz ./torch-xpu-ops/.ci/benchmarks/ ./benchmarks/dynamo/

# optional for torchbench models only
git clone https://github.com/pytorch/benchmark tmp-benchmark
cd tmp-benchmark
python install.py --continue_on_fail <modela,modelb,...>
cd ..

# uninstall the triton and install new from https://github.com/pytorch/pytorch/actions/runs/25853155119?pr=182438

# run affected models (inference)
python benchmarks/dynamo/torchbench.py -d xpu --performance --bfloat16 \
  --inference --backend=inductor --cold-start-latency -n10 \
  --disable-cudagraphs --only resnet50

python benchmarks/dynamo/timm_models.py -d xpu --performance --bfloat16 \
  --inference --backend=inductor --cold-start-latency -n10 \
  --disable-cudagraphs --only mobilenetv2_100
```

## Workflow Runs

- Target: https://github.com/intel/torch-xpu-ops/actions/runs/25921168232
- Baseline: https://github.com/intel/torch-xpu-ops/actions/runs/25737207406

## Expected Behavior

Inductor latency should be comparable to or better than baseline for inference workloads (ratio ~1.0x).

**Comment by anmyachev** [comment]
Hi @mengfei25 just to be sure:
```bash
 ~/intel-xpu-backend-for-triton/.scripts_cache/pytorch$ python benchmarks/dynamo/timm_models.py -d xpu --performance --bfloat16 \
  --inference --backend=inductor --cold-start-latency -n10 \
  --disable-cudagraphs --only mobilenetv2_100

model.safetensors: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 14.2M/14.2M [00:02<00:00, 6.16MB/s]
loading model: 0it [00:04, ?it/s]█████████████████████████████████████████████████████████████████████████████████████████| 14.2M/14.2M [00:02<00:00, 6.16MB/s]
xpu  eval  mobilenetv2_100                    
running benchmark: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 31.64it/s]
2.113x
```

Should I focus on `2.113x` number?

**Comment by anmyachev** [comment]
Probably such lines are problem: `[17:59:23] [INFO] [Worker 2] [150/650 23%] PASS  timm_models/fbnetc_100 (float16, inference, performance) (44s) → 0.112x`


I also see such degradations for float32 types: `timm_models/mobilenetv2_100 (float32, inference, performance) (29s) → 0.182x` (so it's not 16 bit types related)

**UPD:** in our environment this does not occur with either LTS2 driver or Rolling one. @dev-tomek could you try exact steps from the reproducer?

**Comment by dev-tomek** [comment]
@anmyachev I cannot reproduce with the provided setup (pvc1550 + lts2, contenerized env according to the ticket description). Getting ~1x for inductor ratio for both mobilenet and resnet.

**Comment by anmyachev** [comment]
@mengfei25 maybe there were some problems with the hardware or the environment? If you rerun some of the tests, for example just TIMM models, will this problem still occur?

**Comment by mengfei25** [comment]
> [@mengfei25](https://github.com/mengfei25) maybe there were some problems with the hardware or the environment? If you rerun some of the tests, for example just TIMM models, will this problem still occur?
@anmyachev  Still regreesion in my side.
<img width="984" height="715" alt="Image" src="https://github.com/user-attachments/assets/f4a8304b-a293-423c-9746-217f3302730f" />

**Comment by dev-tomek** [comment]
ok managed to get the reproduction, bisection result points to this commit [a559a20aa](https://github.com/intel/intel-xpu-backend-for-triton/commit/a559a20aab5cc1ad442773b4a670cf1cb7110edb)
need to evaluate it before proposing next steps

**Comment by dev-tomek** [comment]
reading the commit message from [a559a20aa](https://github.com/intel/intel-xpu-backend-for-triton/commit/a559a20aab5cc1ad442773b4a670cf1cb7110edb), looks like this is related https://github.com/intel/intel-xpu-backend-for-triton/issues/6261

cc @whitneywhtsang

**Comment by whitneywhtsang** [comment]
The performance degradation on LTS is expected, because `SPV_INTEL_rounded_divide_sqrt` is not supported on LTS.

**Comment by anmyachev** [comment]
@whitneywhtsang Triton 21033c4e2be that was used with the previous release of Pytorch already includes the following changes and they works for these E2E models (on LTS driver):
```cpp
  patterns.add<OpToExternCallConversion<triton::PreciseSqrtOp>>(
      typeConverter, axisInfoAnalysis, "sqrt_cr", benefit);
  patterns.add<OpToExternCallConversion<triton::PreciseDivFOp>>(
      typeConverter, axisInfoAnalysis, "divide_cr", benefit);
```

It turns out that in most cases they work, how can we leave this code and make a fallback only in the case described in https://github.com/intel/intel-xpu-backend-for-triton/issues/6261?

**Comment by whitneywhtsang** [comment]
Those functions are not officially supported by IGC, they have now introduced an officially supported SPV extension for us to use in rolling drivers.

**Comment by anmyachev** [comment]
> Those functions are not officially supported by IGC, they have now introduced an officially supported SPV extension for us to use in rolling drivers.

I don't see a better way, what is your suggestion?

**Comment by whitneywhtsang** [comment]
> > Those functions are not officially supported by IGC, they have now introduced an officially supported SPV extension for us to use in rolling drivers.
> 
> I don't see a better way, what is your suggestion?

IMO we should accept that we cannot get good performance on LTS, until LTS is updated to support the SPV extension. We can also provide an env var to allow usage of those functions, but that's a workaround.

**Comment by anmyachev** [comment]
> > > Those functions are not officially supported by IGC, they have now introduced an officially supported SPV extension for us to use in rolling drivers.
> > 
> > 
> > I don't see a better way, what is your suggestion?
> 
> IMO we should accept that we cannot get good performance on LTS, until LTS is updated to support the SPV extension. We can also provide an env var to allow usage of those functions, but that's a workaround.

Let's discuss it offline.

**Comment by anmyachev** [comment]
@mengfei25 the fix is in release branch already: [753f818](https://github.com/intel/intel-xpu-backend-for-triton/commit/753f818a9375ef5f8ef7ba071dba4a16a9888bf2) and we also have updated Triton in https://github.com/pytorch/pytorch/pull/182438. Could you double check it on your side?

**Comment by anmyachev** [comment]
Fixed in #7006
